toolchain/sem_ir/clang_decl.h

```diff
@@ -5,61 +5,92 @@
-//
-// This type is specifically designed for use in a `CanonicalValueStore` and
-// provide a single canonical access from SemIR to each `clang::Decl*` used.
-// This also ensures that a given `clang::Decl*` is associated with exactly one
-// instruction, and the `inst_id` here provides access to that instruction from
-// either the `ClangDeclId` or the `clang::Decl*`.
-struct ClangDecl : public Printable<ClangDecl> {
+// A canonical declaration pointer is used so that we can perform direct address
+// comparisons and hash this structure based on its contents.
+struct ClangDeclKey : public Printable<ClangDeclKey> {
+  // For declaration classes that are unrelated to FunctionDecl, no parameter
+  // count is expected.
+  template <typename DeclT>
+    requires(std::derived_from<DeclT, clang::Decl> &&
+             !std::derived_from<clang::FunctionDecl, DeclT> &&
+             !std::derived_from<DeclT, clang::FunctionDecl>)
+  explicit ClangDeclKey(DeclT* decl) : decl(decl->getCanonicalDecl()) {}
+
+  // For declaration classes that are derived from FunctionDecl, a parameter
+  // count is required.
+  explicit ClangDeclKey(clang::FunctionDecl* decl, int params)
+      : decl(decl->getCanonicalDecl()), params(params) {}
+
+  // Factory function for clang declaration that is dynamically known to not be
+  // a function declaration.
+  static auto NonFunctionDecl(clang::Decl* decl) -> ClangDeclKey {
+    CARBON_CHECK(!isa<clang::FunctionDecl>(decl));
+    return ClangDeclKey(decl, UncheckedTag());
+  }
+
   auto Print(llvm::raw_ostream& out) const -> void;
 
-  // Equality comparison uses the address-identity property of the Clang AST and
-  // just compares the `decl` pointers. The `inst_id` is always the same due to
-  // the canonicalization.
-  auto operator==(const ClangDecl& rhs) const -> bool {
-    return decl == rhs.decl;
-  }
-  // Support direct comparison with the Clang AST node pointer.
-  auto operator==(const clang::Decl* rhs_decl) const -> bool {
-    return decl == rhs_decl;
+  auto PrintFields(llvm::raw_ostream& out) const -> void;
+
+  auto operator==(const ClangDeclKey& rhs) const -> bool {
+    return decl == rhs.decl && params == rhs.params;
   }
 
   // Hashing for ClangDecl. See common/hashing.h.
-  friend auto CarbonHashValue(const ClangDecl& value, uint64_t seed)
+  friend auto CarbonHashValue(const ClangDeclKey& value, uint64_t seed)
       -> HashCode {
-    return HashValue(value.decl, seed);
+    // Manual hashing support is required because this type has tail padding in
+    // 64-bit compilations.
+    return HashValue(std::pair{value.decl, value.params}, seed);
   }
 
   // The Clang declaration pointing to the Clang AST.
   // TODO: Ensure we can easily serialize/deserialize this. Consider
   // `clang::LazyDeclPtr`.
   clang::Decl* decl = nullptr;
 
+  // The number of parameters to import for a function declaration. Always -1
+  // for a non-function declaration.
+  int params = -1;
```

Comment:
Should we be explicit about size, since this is part of storage?

```suggestion
  int32_t params = -1;
```

Perhaps also `num_` to make it clearer at uses that it's a count, not a reference to actual params? (you also use `num_params` a few spots in this PR already, particularly `CalleeFunctionInfo`)

```suggestion
  int32_t num_params = -1;
```

Reply:
Done.